# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1466040195 25200
#      Wed Jun 15 18:23:15 2016 -0700
# Node ID f500cde065f7c5e00829fb167b2a7990cdf2977f
# Parent  3a0c90b59584958c6f603fa6dd01603aa5a4e0f1
uicontrol: Escape '&' characters in Qt implementation (bug #48214)

* ButtonControl.cc (ButtonControl::ButtonControl, ButtonControl::update):
Escape all '&' characters in "string" property so they are not used by Qt
to create a keyboard shortcut.

diff --git a/libgui/graphics/ButtonControl.cc b/libgui/graphics/ButtonControl.cc
--- a/libgui/graphics/ButtonControl.cc
+++ b/libgui/graphics/ButtonControl.cc
@@ -34,17 +34,19 @@ along with Octave; see the file COPYING.
 namespace QtHandles
 {
 
 ButtonControl::ButtonControl (const graphics_object& go, QAbstractButton* btn)
   : BaseControl (go, btn), m_blockCallback (false)
 {
   uicontrol::properties& up = properties<uicontrol> ();
 
-  btn->setText (Utils::fromStdString (up.get_string_string ()));
+  QString str = Utils::fromStdString (up.get_string_string ());
+  str.replace ("&", "&&");
+  btn->setText (str);
   if (btn->isCheckable () || up.style_is ("togglebutton"))
     {
       btn->setCheckable (true);
 
       Matrix value = up.get_value ().matrix_value ();
 
       if (value.numel () > 0 && value(0) == up.get_max ())
         btn->setChecked (true);
@@ -62,18 +64,22 @@ void
 ButtonControl::update (int pId)
 {
   uicontrol::properties& up = properties<uicontrol> ();
   QAbstractButton* btn = qWidget<QAbstractButton> ();
 
   switch (pId)
     {
     case uicontrol::properties::ID_STRING:
-      btn->setText (Utils::fromStdString (up.get_string_string ()));
-      break;
+      {
+        QString str = Utils::fromStdString (up.get_string_string ());
+        str.replace ("&", "&&");
+        btn->setText (str);
+        break;
+      }
 
     case uicontrol::properties::ID_VALUE:
       m_blockCallback = true;
       if (btn->isCheckable ())
         {
           Matrix value = up.get_value ().matrix_value ();
 
           if (value.numel () > 0)
